pyracf/common/utilities/logger.py

```diff
@@ -149,30 +149,97 @@ def __redact_request_dictionary(
+        trait_key: str,
+    ) -> Union[str, bytes]:
+        r"""
+        Redacts characters in a string between a starting index and ending tag.
         Replaces the identified characters with '********' regardless of the original length.
+
+        This function employs the following regular expressions explained below
+
+        Regex 1 ("quoted") - {trait_key.upper()}( +){{0,}}\(\'.*?(?<!\')\'\)
+        This is designed to match the pattern TRAIT('value') by matching the TRAIT name, a
+        variable (potentially zero) amount of spaces, then the ('value') portion which must
+        start and end with (' and '), but can conceivably contain any characters, but a negative
+        lookbehind is used to look for any unescaped single quotes that would indicate the matching
+        of ') is otherwise a coincidence.
+
+        Regex 2 ("nested") - {trait_key.upper()}( +){{0,}}\([^)]*\(.*\)( +){{0,}}\)
+        This is designed to match the pattern TRAIT( subtrait1(value) subtrait2(value)) by
+        matching the TRAIT name, a variable (potentially zero) amount of spaces, then the
+        ( subtrait1(value) subtrait2(value)) portion which must start and end with ( and ),
+        but must also contain a nested set of opened parentheses rather than a direct seqence of
+        them. The pattern looks for these nested open parenthesis as a sequence would have a )
+        character between them. Then the expression allows any non-newline characters and the
+        "end pattern" of ) and ) separated by a variable (potentially zero) whitespace.
+
+        If neither of these two patterns is found for a supplied trait_key, then it is assumed
+        this trait is set with the default pattern below.
+
+        Regex 3 ("default") - {trait_key.upper()}( +){{0,}}\(.*?(?<!\\)\)
+        This is designed to match the pattern TRAIT(value) by matching the TRAIT name, a variable
+        (potentially zero) amount of spaces, then the (value) portion which must start and end
+        with ( and ), but can conceivably contain any characters, but a negative lookbehind
+        is used to look for any escape \ character that would indicate the matching of the
+        ( and ) is otherwise a coincidence.
+
+        In all replacement expressions, the variable amounts of whitespace are captured so that
+        they can be preserved by this redaction operations.
+
         """
         asterisks = "********"
+        is_bytes = False
         if isinstance(input_string, bytes):
-            asterisks = asterisks.encode("cp1047")
-        pre_keyword = input_string[:start_ind]
-        post_keyword = end_pattern.join(input_string[start_ind:].split(end_pattern)[1:])
-        return pre_keyword + asterisks + end_pattern + post_keyword
+            input_string = input_string.decode("cp1047")
+            is_bytes = True
+        quoted = re.search(
+            rf"{trait_key.upper()}( +){{0,}}\(\'.*?(?<!\')\'\)", input_string
```

Comment:
Some of the regexes are repeated multiple times. It might make sense to initialize these regexes as variables at the beginning. This will not only decrease repetition, which reduces the likelihood that any instance of a repeated value has a typo that is more likely to get overlooked due to duplication, but it will also increase the signal to noise ratio of the code that follows making it easier to read and understand.

```python
asterisks = "********"
is_bytes = False
quoted_regex = rf"{trait_key.upper()}( +){{0,}}\(\'.*?(?<!\')\'\)"
nested_regex = rf"{trait_key.upper()}( +){{0,}}\([^)]*\(.*\)( +){{0,}}\)"
default_regex = rf"{trait_key.upper()}( +){{0,}}\(.*?(?<!\\)\)"
redacted_regex = rf"{trait_key.upper()}\1({asterisks})"
```

Reply:
Excellent points, I had to add a regex for `redacted_w_quotes`, but otherwise adopted this.